src/app/binding/BindingTable.cpp

```diff
@@ -0,0 +1,237 @@
+    mBindingEntries.ForEachActiveObject([&foundEntry, peer](BindingTableEntry * entry) {
+        if (entry->mPeerId == peer)
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(EndpointId endpoint, ClusterId cluster)
+{
+    BindingTableEntry * foundEntry = nullptr;
+    mBindingEntries.ForEachActiveObject([&foundEntry, endpoint, cluster](BindingTableEntry * entry) {
+        if (entry->mSourceEndpoint == endpoint && entry->mClusterId == cluster)
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+OperationalDeviceProxy * BindingTable::FindPeerDevice(NodeId node, FabricIndex fabricIndex)
+{
+    return FindPeerDevice(BuildPeerId(node, fabricIndex));
+}
+
+OperationalDeviceProxy * BindingTable::FindPeerDevice(PeerId peer)
+{
+    OperationalDeviceProxy * foundDevice = nullptr;
+    mPeerDevices.ForEachActiveObject([&foundDevice, peer](OperationalDeviceProxy * device) {
+        if (device->GetPeerId() == peer)
+        {
+            foundDevice = device;
+            return false;
+        }
+        return true;
+    });
+    return foundDevice;
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData)
+{
+    OperationalDeviceProxy * device = mBindingTable->FindPeerDevice(nodeData.mPeerId);
+    if (nodeData.mNumIPs > 0 && device != nullptr && !device->IsConnected() && !device->IsConnecting())
+    {
+        ChipLogProgress(AppServer, "BindingTable: NodeId resolved for 0xdevice %" PRIX64, nodeData.mPeerId.GetNodeId());
+        Transport::PeerAddress addr(nodeData.mAddress[0], Transport::Type::kUdp);
+        addr.SetInterface(nodeData.mInterfaceId);
+        device->UpdateDeviceData(addr, nodeData.mMrpRetryIntervalIdle, nodeData.mMrpRetryIntervalActive);
+        device->Connect(&mBindingTable->mOnDeviceConnected, &mBindingTable->mOnDeviceConnectionFailure);
+    }
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error)
+{
+    ChipLogError(AppServer, "Failed to resolve device 0x%" PRIX64, peerId.GetNodeId());
+}
+
+void BindingTable::OnDeviceConnectedFn(void * context, chip::DeviceProxy * device)
+{
+    ChipLogProgress(AppServer, "Connected to device 0x%" PRIX64, device->GetDeviceId());
+}
+
+void BindingTable::OnDeviceConnectionFailureFn(void * context, NodeId deviceId, CHIP_ERROR error)
+{
+    ChipLogError(AppServer, "Failed to connect to device 0x%" PRIX64, deviceId);
+}
+
+void MatterBindingAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t mask, uint8_t type,
+                                          uint16_t size, uint8_t * value)
+{
+    BindingTableEntry * entry = BindingTable::GetInstance().FindBindingEntry(attributePath.mEndpointId, attributePath.mClusterId);
```

Comment:
OK, so this is wrong not least because it doesn't look at all entries that match the endpoint/cluster.

But also, this is basically assuming that the time to do something with bindings is when one of our attributes changes?  I guess for on/off that might be enough, ok.

Reply:
> when one of our attributes changes

But...we're on a client, not a server? There are no on/off server attributes on a client right?